qiskit_nature/second_q/operators/polynomial_tensor.py

```diff
@@ -0,0 +1,119 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Polynomial Tensor class"""
+
+from __future__ import annotations
+from typing import Dict
+from numbers import Number
+import numpy as np
+from qiskit.quantum_info.operators.mixins import (
+    LinearMixin,
+    AdjointMixin,
+    TolerancesMixin,
+)
+
+
+class PolynomialTensor(LinearMixin, AdjointMixin, TolerancesMixin):
+    """Polynomial Tensor class"""
+
+    def __init__(self, data: Dict[str, np.ndarray]):
+        if not isinstance(data, Dict):
+            raise TypeError(f"data {data} must be a Dictionary")
+
+        self._data = data
+
+        shapes = ()
+        for key, value in self._data.items():
+            shapes += value.shape
+            if len(value.shape) != len(key):
+                raise ValueError(
+                    f"data key {key} of length {len(key)} does not match "
+                    f"data value matrix of dimensions {value.shape}"
+                )
+            if len(set(value.shape)) != 1:
+                raise ValueError(
+                    f"For key {key}: dimensions of value matrix are not identical {value.shape}"
+                )
+
+        if len(set(shapes)) != 1:
+            raise ValueError("Dimensions of value matrices in data dictionary are not identical.")
```

Comment:
How about the following instead?
```suggestion
        shapes = set()
        for key, value in self._data.items():
            if len(value.shape) != len(key):
                raise ValueError(
                    f"data key {key} of length {len(key)} does not match "
                    f"data value matrix of dimensions {value.shape}"
                )

            shape = set(value.shape)

            if len(shape) != 1:
                raise ValueError(
                    f"For key {key}: dimensions of value matrix are not identical {value.shape}"
                )

            shapes.update(shape)

        if len(shapes) != 1:
            raise ValueError("Dimensions of value matrices in data dictionary are not identical.")
```
This has the following improvements:
- only adds `value.shape` to `shapes`, if the checks pass
- uses `shapes` as a `set` directly, without going through a `tuple` first